fem/quadinterpolator_eval_by_vdim.cpp

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "quadinterpolator.hpp"
+#include "quadinterpolator_eval.hpp"
+#include "../general/forall.hpp"
+#include "../linalg/dtensor.hpp"
+#include "../linalg/kernels.hpp"
+
+namespace mfem
+{
+
+template<>
+void QuadratureInterpolator::Values<QVectorLayout::byVDIM>(
+   const Vector &e_vec, Vector &q_val) const
+{
+   const int NE = fespace->GetNE();
+   if (NE == 0) { return; }
+   const int vdim = fespace->GetVDim();
+   const int dim = fespace->GetMesh()->Dimension();
+   const FiniteElement *fe = fespace->GetFE(0);
+   const IntegrationRule *ir =
+      IntRule ? IntRule : &qspace->GetElementIntRule(0);
+   const DofToQuad &maps = fe->GetDofToQuad(*ir, DofToQuad::TENSOR);
+   const int D1D = maps.ndof;
+   const int Q1D = maps.nqpt;
+   const double *B = maps.B.Read();
+   const double *X = e_vec.Read();
+   double *Y = q_val.Write();
+
+   constexpr QVectorLayout L = QVectorLayout::byVDIM;
+
+   const int id = (dim<<12) | (vdim<<8) | (D1D<<4) | Q1D;
+
+   switch (id)
+   {
+      case 0x2124: return Eval2D<L,1,2,4,8>(NE,B,X,Y);
+      case 0x2136: return Eval2D<L,1,3,6,4>(NE,B,X,Y);
+      case 0x2148: return Eval2D<L,1,4,8,2>(NE,B,X,Y);
+
+      case 0x2224: return Eval2D<L,2,2,4,8>(NE,B,X,Y);
+      case 0x2234: return Eval2D<L,2,3,4,8>(NE,B,X,Y);
+      case 0x2236: return Eval2D<L,2,3,6,4>(NE,B,X,Y);
+      case 0x2248: return Eval2D<L,2,4,8,2>(NE,B,X,Y);
+
+      case 0x3124: return Eval3D<L,1,2,4>(NE,B,X,Y);
+      case 0x3136: return Eval3D<L,1,3,6>(NE,B,X,Y);
+      case 0x3148: return Eval3D<L,1,4,8>(NE,B,X,Y);
+
+      case 0x3324: return Eval3D<L,3,2,4>(NE,B,X,Y);
+      case 0x3336: return Eval3D<L,3,3,6>(NE,B,X,Y);
+      case 0x3348: return Eval3D<L,3,4,8>(NE,B,X,Y);
+
```

Comment:
Why does the `byVDIM` version have fewer `id` cases supported than `byNODES`? Also, can't the two versions be combined into one piece code rather than two?

Reply:
By default, most of the code uses `byNODES`, but some others like the `sedov` unit test, CEED's `Laghos` and the production code start to use the `byVDIM` layout, which allows to manipulate one pointer to the data. I populated the compiled-time kernels when I encountered them.
It uses the unique code of the header, and splitting allows to break the compilation time for the two kernels. I'm fine either way, but one goal was to try to reduce this time.

Reply:
Reducing compilation time is a good enough reason to keep it the way it is.